File: backend/app/models/anomaly.py
import uuid

from sqlalchemy import ForeignKey, Numeric, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, TIMESTAMP, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database.session import Base


class AnomalyFlag(Base):
    __tablename__ = "anomaly_flags"
    __table_args__ = (
        UniqueConstraint(
            "household_id", "rule", "dedupe_key", name="uq_anomaly_flags_identity"
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    household_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("households.id"), nullable=False
    )
    transaction_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("transactions.id"), nullable=True
    )
    rule: Mapped[str] = mapped_column(String, nullable=False)
    dedupe_key: Mapped[str] = mapped_column(String, nullable=False)
    severity: Mapped[str] = mapped_column(String, nullable=False)
    score: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    summary: Mapped[str] = mapped_column(String, nullable=False)
    raw_context: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    status: Mapped[str] = mapped_column(String, nullable=False, default="open")
    explanation: Mapped[str | None] = mapped_column(String, nullable=True)
    explained_at: Mapped[TIMESTAMP | None] = mapped_column(
        TIMESTAMP(timezone=True), nullable=True
    )
    created_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

File: frontend/lib/data/models/anomaly.dart
class AnomalySummary {
  const AnomalySummary({
    required this.id,
    required this.transactionId,
    required this.householdMemberId,
    required this.rule,
    required this.severity,
    required this.score,
    required this.summary,
    required this.status,
    required this.explanation,
    required this.explainedAt,
    required this.createdAt,
  });

  factory AnomalySummary.fromJson(Map<String, dynamic> json) {
    return AnomalySummary(
      id: json['id'] as String,
      transactionId: json['transaction_id'] as String?,
      householdMemberId: json['household_member_id'] as String?,
      rule: json['rule'] as String,
      severity: json['severity'] as String,
      score: (json['score'] as num?)?.toDouble(),
      summary: json['summary'] as String,
      status: json['status'] as String,
      explanation: json['explanation'] as String?,
      explainedAt: json['explained_at'] == null
          ? null
          : DateTime.parse(json['explained_at'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  final String id;
  final String? transactionId;
  final String? householdMemberId;
  final String rule;
  final String severity;
  final double? score;
  final String summary;
  final String status;
  final String? explanation;
  final DateTime? explainedAt;
  final DateTime createdAt;

  AnomalySummary copyWith({
    String? status,
    String? explanation,
    DateTime? explainedAt,
  }) {
    return AnomalySummary(
      id: id,
      transactionId: transactionId,
      householdMemberId: householdMemberId,
      rule: rule,
      severity: severity,
      score: score,
      summary: summary,
      status: status ?? this.status,
      explanation: explanation ?? this.explanation,
      explainedAt: explainedAt ?? this.explainedAt,
      createdAt: createdAt,
    );
  }
}
